### ISSUE TITLE
[feature request]New pytorch op for something like combinatoric iterators in itertools

### ISSUE THREAD

**Issue by zasdfgbnm** [pytorch/pytorch#7580]
Hi, I want some ops that is similar to [combinatoric iterators in itertools](https://docs.python.org/3.6/library/itertools.html). These functions do the same thing as in itertools. For example:
```python
>>> A = torch.tensor([1,2,3])
>>> B = torch.tensor([4,5,6])
>>> cartesian_prod(A,B)
tensor([[ 1,  1,  1,  2,  2,  2,  3,  3,  3],
        [ 4,  5,  6,  4,  5,  6,  4,  5,  6]])
>>> combinations(A, 2)
tensor([[ 1,  1,  2],
        [ 2,  3,  3]])
>>> combinations_with_replacement(A, 2)
tensor([[ 1,  1,  1,  2,  2,  3],
        [ 1,  2,  3,  2,  3,  3]])
```
I already have [a python implementation of these ops](https://gist.github.com/zasdfgbnm/dfbe1c29ce8073dc2224c7286583fa43), supporting tensors of arbitrary rank, implemented using `torch.unbind`, `torch.stack`, and `torch.cat`. But this implementation is very slow, and is a bottleneck of my application.

So I'm just asking, is pytorch interested in supporting these natively? If yes, I'm happy to contribute.

**Comment by zasdfgbnm** [comment]
@fmassa Yes, the above example can be done easily by the following code:
```python
>>> _A = A.view(-1,1).expand(3,3).reshape(-1)
>>> _B = B.repeat(3)
>>> torch.stack([_A, _B])
tensor([[ 1,  1,  1,  2,  2,  2,  3,  3,  3],
        [ 4,  5,  6,  4,  5,  6,  4,  5,  6]])
```
But if you have a high rank tensor, say (100, 100000, 10), and you want to do Cartesian product on `dim=1`, It requires much more effort on doing so.

Also, I don't find a way to do `combinations` efficiently. The thing is, if the pytorch team don't think these ops are helpful enough to add it to pytorch, then there is nothing to worry about; but if the pytorch team do think these ops are helpful, since `combinations` and `combinations_with_replacement` are doing very similar job as `cartesian_prod`, then it might be better to make `cartesian_prod` come together rather than force the user to implement it by themselves.

**Comment by fmassa** [comment]
Quick question: what's the advantage of viewing the tensor as 1d and concatenating them? Is it to make sampling easier? For sampling as if it was a 1d tensor given a list of indices (say randpermon the numel of the meshgrid), one could use `torch.take` and I believe the results would be fairly efficient (but need to be performed twice, one for x and one for y, the results of meshgrid).
But I might be missing something in here.
Would be good to know what the others think

**Comment by zasdfgbnm** [comment]
@fmassa In my application, it is not for sampling. I'm dealing with molecules and atoms. Without worrying about details of my applications, you can think this way:

In one batch, there are 1000 frames. For each frame, there are 100 points. The input is tensor of shape (1000, 100, 3) storing the xyz coordinates of all the points. And I need to compute some functions (for all frames in batch) based on the distances and angles of points within each frame. This is basically the reason why I need something to do combinatoric iterators along a dimension(here dim=1).

Hope this explanation is clear.

**Comment by zasdfgbnm** [comment]
@fmassa Another application might be (not sure), for example if you have a tensor of shape (128, 512, 512) storing 128 images each of size 512*512 in gray scale. Then doing a `combinations_with_replacement` would get a tensor of shape (8256, 512, 512, 2) of image pairs 8256 possible pairs of images. The resulting tensor might be useful for batch processing for models that takes pairs of images as input. These are not my area of study, so I don't know if the example above is good or not.

**Comment by fmassa** [comment]
@zasdfgbnm I think I understand the use-case. I misused the work "sampling" to express what I meant earlier on.
Maybe a small (and untested) code snippet can better express what I meant:
```python
# helper function
def meshgrid(x, y=None):
    if y is None:
        y = x
    x = torch.as_tensor(x)
    y = torch.as_tensor(y)
    m, n = x.size(0), y.size(0)
    grid_x = x[None].expand(n, m)
    grid_y = y[:, None].expand(n, m)
    return grid_x, grid_y

def combinations(x, y, dim, n):
    """
    Given tensors x and y, get a random combination
    of the elements of x and y over dimension dim
    """
    xs = torch.arange(x.shape[dim])
    ys = torch.arange(y.shape[dim])
    grid_x, grid_y = meshgrid(xs, ys)
    # select n random elements from the
    # cartesian product
    sampled = torch.randperm(grid_x.numel())[:n]
    indices_x = grid_x.take(sampled)
    indices_y = grid_y.take(sampled)
    # get from the indices
    return x.index_select(dim, indices_x), y.index_select(dim, indices_y) 
```

There might be value in adding something like that in pytorch, but as I said before, it might be good to have the opinion from the others from the team.

**Comment by zasdfgbnm** [comment]
@fmassa Ahh, I understand what you mean now. And yes, let's wait for other people's opinion.

**Comment by apaszke** [comment]
I don't mind adding those too much, but I'm not sure how useful they are. Cartesian products, combinations generally cause a significant blow up in object sizes, and considering that every element has already been an array, it will be super easy to cause an OOM error because of that. It seems better to me to lazily create those elements (possibly batch them) and process them in this order.

What's more, we won't get any specialized kernels for them, because they're generally implementing using 2-4 indexing ops, so that would be an overkill (it's basically a strided copy). That would be a lot of work for a tiny speedup.

**Comment by zasdfgbnm** [comment]
@fmassa @apaszke I think maybe we can implement something like:
```
>>> torch.cartesian_range(3)
tensor([[ 0,  0,  0,  1,  1,  1,  2,  2,  2],
        [ 0,  1,  2,  0,  1,  2,  0,  1,  2]])
>>> torch.combination_range(3)
tensor([[ 0,  0,  1],
        [ 1,  2,  2]])
>>> torch.combination_range_with_replacement(3)
tensor([[ 0,  0,  0,  1,  1,  2],
        [ 0,  1,  2,  1,  2,  2]])
```
Then the full product can easily be implemented using `index_select`

**Comment by zasdfgbnm** [comment]
@fmassa Ahh, yes. So, do you think there is no need to implement a `cartesian_range`? To be honest, if I can only choose one of `meshgrid` and `cartesian_range`, I would prefer `meshgrid`. Do you think it is better to implement a meshgrid into ATen, or leave it for the user to implement themselves?

By the way, I can not think of an application when meshgrid would be helpful(except the example you gave above), because broadcasting semantics itself is working already very well: if you want to compute x + y for all pairs (x,y), just x.view(-1,1) + y.view(1,-1) would work perfectly.

**Comment by zasdfgbnm** [comment]
@fmassa Also, the first index list of `combination_range`, and `combination_range_with_replacement `, i.e. the `[ 0,  0,  1]` and `[ 0,  0,  0,  1,  1,  2]`, can be easily generated by `numpy.repeat`, for example:
```python
>>> a = numpy.arange(3)
>>> repeat = numpy.arange(3,0,-1)
>>> a.repeat(repeat)
array([0, 0, 0, 1, 1, 2])
```
But I don't think pytorch has similar thing, right?
So, would it be helpful if we extend `torch.repeat` to support numpy like API? This sounds more useful than `combination_range` and `combination_range_with_replacement`.

But I don't have an idea on how to implement the second index list yet. Maybe using `torch.triu` as discussed [at the forum](https://discuss.pytorch.org/t/upper-triangular-matrix-vectorization/7040/4)

**Comment by zasdfgbnm** [comment]
If you would like meshgrid to appear in ATen, I'm more than happy to add that. Because it is kind of a standard operator (it appears in matlab, numpy, tensorflow, etc.) and supporting it definitely make it easier for users.

And, I just think of a way to create an index for `combination_range`, and `combination_range_with_replacement` without using repeat. Using [the method of vectorizing upper triangles of a matrix as discuss in the forum](https://discuss.pytorch.org/t/upper-triangular-matrix-vectorization/7040/4).

Basically, `[0,0,0,1,1,2]` is the upper triangle (including diagonal) of
```
[[ 0, 0, 0],
 [ 1, 1, 1],
 [ 2, 2, 2]]
```
And `[0,1,2,1,2,2]` is the upper triangle (including diagonal) of:
```
[[ 0, 1, 2],
 [ 0, 1, 2],
 [ 0, 1, 2]]
```

And yes, I do agree with your statement:

> I'm not seeing many use-cases of repeat that couldn't be achieved with advanced indexing (even though creating the indices would be slightly tricky, but doable without issues in python)

Even though advanced indexing is powerful, creating the indices is highly non-trivial for applications. 
I would consider `combination_range`, and `combination_range_with_replacement` as two examples of this difficulty. And as a general tool, it would be helpful to add function to create indices based on applications simply for the ease of use, even though some of these functions can be easily implemented by some trick. What `combination_range`, and `combination_range_with_replacement` trying to do is to provide a simple way of creating things like that.

**Comment by zasdfgbnm** [comment]
Another (probably better) option, instead of adding `cartesian_range`, `combination_range`, and `combination_range_with_replacement` as an op of pytorch, we can just implement `meshgrid`, and in the doc of meshgrid, we gives 3 examples of application of using meshgrid to implement cartesian product, combinations, and combination with replacement.

**Comment by OhadRubin** [comment]
if someone is looking for a batched version:
`def cartesian_product(tensor):
    D = tensor.reshape(-1,1)
    _B = torch.stack([x.repeat(x.size(0)) for x in tensor]).reshape(-1)
    _D = D.view(-1,1).expand(tensor.size(0)*tensor.size(1),tensor.size(1)).reshape(-1)
    return torch.stack([_D, _B]).T.reshape(tensor.size(0),tensor.size(1)**2,2)
d=10
A = torch.arange(d)
C = -torch.arange(d)
cartesian_product(torch.stack([A,C]))`
